### ISSUE TITLE
Unable to correctly link two deep level arguments sharing the same root class

### ISSUE THREAD

**Issue by AmirSa7** [omni-us/jsonargparse#297]
I've been working on trying to make this code function in a clean and logical manner, but despite multiple attempts, I'm uncertain whether it's feasible or not, and could not find a similar example online.
I've come across a few issues related to similar challenges, but I haven't been able to achieve the desired behavior with my code.

Thanks!

**To Reproduce**
Use this `config.yaml`:
```yaml
model:
  class_path: __main__.LitDetector
  init_args:
    detector:
      class_path: __main__.DetectionModel
      init_args:
        encoder:
          class_path: __main__.Encoder
          init_args:
            model_name: modelA
        decoder:
          class_path: __main__.Decoder
          init_args:
            output_channels: 256
        det_const: 10
    lit_const: 12
```

and this code:
```python
from jsonargparse import CLI
from jsonargparse import ArgumentParser

class Encoder():
    def __init__(self, model_name: str):
        self.output_channels = dict(modelA=16,
                                    modelB=32)[model_name]

class Decoder():
    def __init__(self,
                 input_channels: int,
                 output_channels: int = 256):
        assert input_channels is not None, "Linking Failed"
        self.input_channels = input_channels
        self.output_channels = output_channels

class DetectionModel():
    def __init__(self,
                 encoder: Encoder,
                 decoder: Decoder,
                 det_const: int = 10) -> None:
        self.encoder = encoder
        self.decoder = decoder
        self.det_const = det_const

class LitDetector:
    def __init__(self,
                 detector: DetectionModel,
                 lit_const: int = 12) -> None:
        self.detector = detector
        self.lit_const = lit_const

class DetectorCLI:
    def __init__(self, model: LitDetector) -> None:
        self.model = model
        self.data = None
        print('>>> Done.')

if __name__ == '__main__':
    parser = ArgumentParser(prog='nested_classes_example',
                            description='Example of nested classes')
    parser.link_arguments('model.detector.encoder.output_channels',
                          'model.init_args.detector.init_args.decoder.init_args.input_channels',
                          apply_on='instantiate')
    args = parser.parse_args()
    CLI(DetectorCLI)
```

**Expected Behavior**
The goal is to successfully init `LitDetector`. For that we need to init `Encoder` and `Decoder` while the second is dependent in the first. Enough information is supplied in order to init the first. After the `Encoder` had been initiated we can deduce the relevant information for the initiation of the `Decoder`, using `link_arguments`.

For the clarity, In most attempts it fails before executing the "assert" line.

**Environment**
- jsonargparse version: 4.21.2
- Python version: 3.11.3
- How jsonargparse was installed: pip3 install jsonargparse[signatures]
- OS: Mac

**Comment by mauvilsa** [comment]
From a quick look at it, there seems to be a bug in jsonargparse. When I understand better what is happening or have a solution, I will comment here. For the moment just one clarification. The `CLI` function internally creates a parser, parses and runs code. Also `CLI` does not support linking of arguments. In the script above the `link_arguments` fails because `model` has not been added. A more appropriate reproduction script would be:

```python
parser = ArgumentParser()
parser.add_argument('--config', action=ActionConfigFile)
parser.add_class_arguments(DetectorCLI)
parser.link_arguments(
    'model.detector.encoder.output_channels',
    'model.init_args.detector.init_args.decoder.init_args.input_channels',
    apply_on='instantiate',
)
cfg = parser.parse_args()
init = parser.instantiate_classes(cfg)
```

**Comment by mauvilsa** [comment]
@AmirSa7 this was fixed in pull request #412. It is now in the main branch and will be included in the next release. Thank you for reporting! Sorry that it took so long to fix.
